File: solutions/typescript/custom-set/1/custom-set.ts
export class CustomSet {
  private elements: Set<unknown>;

  constructor(initial?: unknown[]) {
    // Si viene un arreglo inicial, lo cargamos en el Set nativo para limpiar duplicados
    this.elements = new Set(initial || []);
  }

  // Comprueba si el conjunto está vacío
  public empty(): boolean {
    return this.elements.size === 0;
  }

  // Verifica si un elemento específico pertenece al conjunto
  public contains(element: unknown): boolean {
    return this.elements.has(element);
  }

  // Agrega un elemento único y retorna la propia instancia del conjunto (Method Chaining)
  public add(element: unknown): CustomSet {
    this.elements.add(element);
    return this;
  }

  // Determina si el conjunto actual es un subconjunto de 'other'
  public subset(other: CustomSet): boolean {
    for (const elem of this.elements) {
      if (!other.contains(elem)) {
        return false;
      }
    }
    return true;
  }

  // Determina si el conjunto actual y 'other' no comparten ningún elemento (son disjuntos)
  public disjoint(other: CustomSet): boolean {
    for (const elem of this.elements) {
      if (other.contains(elem)) {
        return false; // Al primer elemento compartido, ya no son disjuntos
      }
    }
    return true;
  }

  // Comprueba si dos conjuntos contienen exactamente los mismos elementos
  public eql(other: CustomSet): boolean {
    if (this.elements.size !== other.elements.size) {
      return false;
    }
    return this.subset(other);
  }

  // Retorna un nuevo conjunto con la combinación de ambos (A ∪ B)
  public union(other: CustomSet): CustomSet {
    const combined = new CustomSet([...this.elements, ...other.elements]);
    return combined;
  }

  // Retorna un nuevo conjunto con los elementos que tienen en común (A ∩ B)
  public intersection(other: CustomSet): CustomSet {
    const commonElements: unknown[] = [];
    for (const elem of this.elements) {
      if (other.contains(elem)) {
        commonElements.push(elem);
      }
    }
    return new CustomSet(commonElements);
  }

  // Retorna un nuevo conjunto con los elementos de A que no pertenecen a B (A - B)
  public difference(other: CustomSet): CustomSet {
    const uniqueElements: unknown[] = [];
    for (const elem of this.elements) {
      if (!other.contains(elem)) {
        uniqueElements.push(elem);
      }
    }
    return new CustomSet(uniqueElements);
  }
}

File: solutions/ruby/custom-set/1/custom_set.rb
class CustomSet
  attr_reader :elements

  def initialize(elements = [])
    # Guardamos los elementos como un arreglo único y limpio de duplicados
    @elements = elements.uniq
  end

  # Comprueba si el conjunto está vacío
  def empty?
    @elements.empty?
  end

  # Verifica la presencia de un elemento específico
  def contains?(element)
    @elements.include?(element)
  end
  
  # Creamos un alias para que los tests que buscan .member? pasen sin problemas
  alias_method :member?, :contains?

  # Determina si el conjunto actual es un subconjunto de otro conjunto
  def subset?(other)
    @elements.all? { |element| other.contains?(element) }
  end

  # Evalúa si dos conjuntos son disjuntos (no tienen elementos en común)
  def disjoint?(other)
    @elements.none? { |element| other.contains?(element) }
  end

  # Sobrescribimos el operador de igualdad
  def ==(other)
    return false unless other.is_a?(CustomSet)
    
    # Dos conjuntos son iguales si tienen el mismo tamaño y todos los elementos coinciden
    @elements.length == other.elements.length && subset?(other)
  end

  # Agrega un elemento mutando el conjunto (y manteniendo la unicidad)
  def add(element)
    @elements << element unless contains?(element)
    self
  end

  # Operación de Intersección: Elementos presentes en ambos conjuntos
  def intersection(other)
    common_elements = @elements.select { |element| other.contains?(element) }
    CustomSet.new(common_elements)
  end

  # Operación de Diferencia: Elementos de este conjunto que NO están en el otro
  def difference(other)
    different_elements = @elements.reject { |element| other.contains?(element) }
    CustomSet.new(different_elements)
  end

  # Operación de Unión: Combina todos los elementos de ambos conjuntos
  def union(other)
    CustomSet.new(@elements + other.elements)
  end
end